Adds ussd send on pin

mango-habanero opened the pull request:
**Describe the Pull Request**

This bug changes the send token flow of the ussd send token process. It removes the need to confirm what a user bought and initiates the sending of the actual token on entry of the user pin.

https://github.com/GrassrootsEconomics/CIC-Docs/projects/1#card-35821324

mango-habanero committed:
Adds ussd send on pin

nolash commented on app/server/locale/ussd.sw.yml:
```diff
@@ -43,16 +43,14 @@ sw:
+        CON Tuma %{transaction_amount} %{token_name} kwa %{recipient_phone}.
+                Weka nambari ya siri kuthibitisha.
+                0. Nyuma
+                0. Nyuma
```
2 x 0?

nolash commented on app/server/utils/ussd/kenya_ussd_processor.py:
```diff
@@ -130,19 +130,25 @@ def custom_display_text(menu: UssdMenu, ussd_session: UssdSession) -> str:
-                transaction_amount=cents_to_dollars(transaction_amount),
-                transaction_reason=transaction_reason
-            )
+            if user.failed_pin_attempts is not None and user.failed_pin_attempts > 0:
```
What should happen at `user.failed_pin_attempts == 0` ? is there code to catch that?

nolash replied:
actually that should be in the test also.

mango-habanero replied:
> What should happen at `user.failed_pin_attempts == 0` ? is there code to catch that?

There's an else there yes ?
If the user.failed_pin_attempts == 0 then the else component in:

```
if user.failed_pin_attempts is not None and user.failed_pin_attempts > 0:
                return i18n_for(
                    user=user,
                    key="{}.{}".format(menu.display_key, 'retry'),
                    remaining_attempts=3 - user.failed_pin_attempts
                )
            else:
                return i18n_for(
                    user=user,
                    key="{}.{}".format(menu.display_key, 'first'),
                    transaction_amount=cents_to_dollars(transaction_amount),
                    token_name=token.symbol,
                    recipient_phone=recipient_phone
                )
```
will run and return the `send_token_pin_authorization` menu's `first` section

mango-habanero replied:
> actually that should be in the test also.

The largest component of this ussd processor is the `custom_display_text` function which is already tested.
So even the scenario you've referred to is already implicitly tested.

nolash replied:
Unless somehow the logical `and` between the two allow a different edge case to fall through...? Why for example isn't it `or` then?

mango-habanero replied:
That's a really good catch. I think its overkill having it as it is. Maybe it should just be `user.failed_pin_attempts > 0:` the field not being none here is implicit

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -131,11 +114,11 @@ def mock_send_message(phone, message):
     org = OrganisationFactory()
     sender = UserFactory(preferred_language="en", phone=make_kenyan_phone(phone()), first_name="Bob", last_name="Foo",
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
```
please args below each other or just one indent

mango-habanero replied:
The [`self-service-wallets-func-tests`](https://github.com/GrassrootsEconomics/SempoBlockchain/blob/philip/self-service-wallets-func-tests/test/functional/app/api/test_ussd_api.py) PR cleans all these lint issues.

nolash commented on test/helpers/ussd_tasker.py:
```diff
@@ -21,4 +26,4 @@ def inquire_balance(user: User):
     @staticmethod
     def fetch_user_exchange_rate(user: User):
-        pass
+        pass
```
do you have some auto-formatting on that messes with the last newline?

mango-habanero replied:
I have pylint running for my scripts.

nolash commented on test/unit/app/utils/ussd/kenya/test_pin_state.py:
```diff
@@ -35,23 +37,23 @@
-     (opt_out_of_market_place_pin_authorization_state, pin_blocked_user, "1111",
-      "exit_pin_blocked")
- ])
+                         [
```
hmm this looks like auto-formatting too. I'm not a fan.

nolash commented on test/unit/app/utils/ussd/kenya/test_pin_state.py:
```diff
@@ -71,12 +73,10 @@ def test_kenya_state_machine(test_client, init_database, user_factory, session_f
-     (send_token_pin_authorization_state, standard_user, "0000",
-      "send_token_confirmation", 1, 0),
+     (send_token_pin_authorization_state, standard_user, "0000", "complete", 1, 0),
      (send_token_pin_authorization_state, standard_user, "1212",
```
maybe all of them on one line then?

nolash commented on test/unit/app/utils/ussd/kenya/test_pin_state.py:
```diff
@@ -110,15 +110,16 @@ def test_kenya_state_machine(test_client, init_database, user_factory, session_f
-    session = session_factory()
-    user = user_factory(phone='+6140000000')
-    user.failed_pin_attempts = before_failed_pin_attempts
+    with db.session.no_autoflush:
```
is there autoflush here?

mango-habanero replied:
Yes there's a query invoked autoflush there

nolash replied:
```
$ grep autoflush test -r
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py:    with db.session.no_autoflush:
Binary file test/unit/app/utils/ussd/kenya/__pycache__/test_kenya_ussd_processor.cpython-36-PYTEST.pyc matches
```

I'm a bit sceptical here. _no other test_ has explicit autoflush check. Why is it needed in _this_ particular case please.

mango-habanero replied:
So when I run the test without the autoflush I run into an error along the lines of there being a user existing who bore the same number as the one that piece of code attempts to create.
The error then suggested that the cause of the error was a premature autoflush that was being invoked by some query within the flow of that piece of code. It the suggested that I run it without the autoflush to avoid the error and for sure it worked.

nolash replied:
the way the other tests work afaiu is that then you use a fixture to create a different user. (look for keyword "other" or "another" or something in the conftest files)

Whereas that's clearly not in any way elegant, I do not think we should introduce session-control methods that aren't introduced before. We are struggling with enough variance as it is in this system. Such can make it harder to clean things up once we make the pass to harden, sanitize and standardize the tests and fixtures.

mango-habanero replied:
So I revisited this today, I don't think there's a way around this when Sqlachemy runs that autoflush thing. 

> ```
> $ grep autoflush test -r
> test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py:    with db.session.no_autoflush:
> Binary file test/unit/app/utils/ussd/kenya/__pycache__/test_kenya_ussd_processor.cpython-36-PYTEST.pyc matches
> ```
> 
> I'm a bit sceptical here. _no other test_ has explicit autoflush check. Why is it needed in _this_ particular case please.

Then I tried to find out if there were situations where it had been the current code base. I found one in:
`test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py` in the `test_custom_display_text` test.
That aside, I am still looking for a way around but it looks like its one of those things where we can pass the no autoflush check in the session wide with:
`autoflush=false` or use what I've used there.

nolash replied:
> . I found one in:
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text 

which commit? I could not find that in your revision.

mango-habanero replied:
> > . I found one in:
> > test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text
> 
> which commit? I could not find that in your revision.

https://github.com/GrassrootsEconomics/SempoBlockchain/commit/8517db39912709f00acf2b5d9d1a9741355c8a00

nolash replied:
```
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 39) def test_custom_display_text(test_client, init_database, menu_name, language, expecteds,
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 40)                              real_at_idx, length, unexpected, menu_nr):
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 41)     with db.session.no_autoflush:
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 42)         start_state = UssdSessionFactory(state="menu_name")
```

Ok, thanks, I yield.

It seems like a hack, though, so we should make a point of weeding this out. I guess we are adding that to technical debt, unless you already found a way to do this cleanly and painlessly? We really need to get this functionality out.

mango-habanero committed:
Clean up tests for circle

mango-habanero committed:
Clean up send on pin

mango-habanero committed:
Merge branch 'master' into philip/send-on-pin

mango-habanero committed:
Merge branch 'master' into philip/send-on-pin

nolash reviewed (commented):
I cannot see other changes than just merge master in...?

PhilipWafula commented:
> I cannot see other changes than just merge master in...?

I don't follow.

> I cannot see other changes than just merge master in...?

I am not sure I follow:

mango-habanero committed:
Makes minor fixes.

- Adds missing 0 for back to main menu on send token pin authorization menu.
- Removes unnecessary check for none.
- Removes the unnecessary string literals on valid service code value.

mango-habanero committed:
Merges master
